lib/FreeRTOS/freertos-plus-mqtt/agent_command_pool.c

```diff
@@ -54,35 +62,54 @@
 static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
 
 /**
- * @brief A counting semaphore used to guard the pool of Command_t structures.  To
- * obtain a structure first decrement the semaphore count.  To return a structure
- * increment the semaphore count after the structure is back in the pool.
+ * @brief A queue used to guard the pool of Command_t structures. Structures
+ * may be obtained by receiving a pointer from the queue, and returned by
+ * sending the pointer back into it.
  */
-static SemaphoreHandle_t freeCommandStructMutex = NULL;
+static AgentMessageContext_t commandStructQueue;
 
-static volatile uint8_t initStatus = SEMAPHORE_NOT_INITIALIZED;
+/**
+ * @brief Initialization status of the queue.
+ */
+static volatile uint8_t initStatus = QUEUE_NOT_INITIALIZED;
 
 /*-----------------------------------------------------------*/
 
 static void initializePool()
 {
     bool owner = false;
+    size_t i;
+    Command_t * pCommand;
+    static uint8_t staticQueueStorageArea[ MQTT_COMMAND_CONTEXTS_POOL_SIZE * sizeof( Command_t * ) ];
+    static StaticQueue_t staticQueueStructure;
 
     taskENTER_CRITICAL();
     {
-        if( initStatus == SEMAPHORE_NOT_INITIALIZED )
+        if( initStatus == QUEUE_NOT_INITIALIZED )
```

Comment:
If there is concurrent access vulnerability of the `initStatus` state variable, should it be atomically accessed within `Agent_GetCommand` for the `if( initStatus == QUEUE_INITIALIZED )` check?

Reply:
We have `taskENTER_CRITICAL` here since we're writing to `initStatus` and `owner`. It's not necessary for the read on line 131. The status doesn't change again after setting to `QUEUE_INITIALIZED`